Novelty Score Calculation Update

pg427 opened the pull request:
Updated novelty score calcualtions including integrations of:
1. Clinical Trials/ Clinical Approvals - MVP1
2.  TDLs - MVP2
3. Gene Distinctiveness - MVP2

pg427 committed:
Novelty Score Calculation Update

Updated novelty score calcualtions including integrations of:
1. Clinical Trials/ Clinical Approvals - MVP1
2.  TDLs - MVP2
3. Gene Distinctiveness - MVP2

maximusunc commented on app/novelty/compute_novelty.py:
```diff
@@ -3,10 +3,27 @@
-from .extr_smile_molpro_by_id import mol_to_smile_molpro
-from .mol_similarity import find_nearest_neighbors
+import json
+from known import find_known_results
+from extr_smile_molpro_by_id import mol_to_smile_molpro
+from mol_similarity import find_nearest_neighbors
```
These three lines are not package imports, but local. They fail to be imported.

pg427 replied:
Got it. I've imported the two necessary functions into the compute_novelty.py file to make it simpler. There won't be a need to import the three files anymore. I'll keep them in the repo in case there is future requirement of use.

maximusunc replied:
If those files are no longer being used, could you remove them please? If needed in the future, we can always reference them in GitHub history. Ideally any commented out code would also be deleted.

pg427 replied:
Understood. I have removed the files from the repo and the commented out code from compute_novelty.py

maximusunc commented on app/novelty/compute_novelty.py:
```diff
@@ -377,98 +463,412 @@ def novelty_score(fda_status, recency, similarity):
+                           'Gene Distinctiveness', 'TDLs', 'novelty_score']
+        df_numpy = pd.DataFrame([["NAN"]*len(column_list)]*len(message['results']), columns = column_list)
+
+    return df_numpy
```
The final output used to only have `drug` and `novelty_score`, with `drug` being the curie. Is that now `Result ID`? Also, at least for the query I'm trying, every value is NaN. If that's expected, could we make the default value be 0 or something?

pg427 replied:
Yes. Since now we have MVP1 and MVP2 queries I changed the 'drug' to 'Result ID'. Can you send me the query? The lowest value is 0 for the novelty score. If it's showing NaN then there might be an issue I'd like to look a bit closer into. I'm assuming the `"message" ` key from the json response is being passed from the merged Response from the ARS.

maximusunc replied:
I'm working with a smaller message, and a large message. The smaller message seems to be fine, but the larger message returns all NaNs. The CI PK for that larger message is d1e6d43f-cb3e-4c6b-a6bc-5393f576053f.

pg427 replied:
I checked the pk and it seems that the message gave an Error code of 422. It is probably why you're seeing ther NaN. I kept the NaN on purpose to distinguish between getting a 0 novelty score and when the code is erroring out. At the moment the code only works for messages with the code of 200 and shows NaNs otherwise. Do you want me to switch these out for 0s ?

maximusunc replied:
The 422 error was given by the ARS, but that's not necessarily indicative of what the Appraiser did. I can pass that message to the Appraiser and it handles it fine, other than the novelty code, so I think something needs to be fixed in this code.

FYI, ALL fields are set to NaN, not just novelty_score.
FYI FYI, if the novelty code errors out, I'm already setting the default novelty to 0.0, so I think it would be a good idea to have that same default value in your code instead of a NaN (which I think is actually the string "NaN" instead of the value NaN.

pg427 replied:
Ok. I've changed the default "NaN" strings (for instances when an error is encountered) to 0 for all entries in the DataFrame generated.

maximusunc replied:
So now the default value is 0, but ALL fields are still 0. I'm also not seeing any error messages come out of novelty, so it looks like everything went fine. In my opinion, `Query ID` and `Result ID` should always have values and not be 0, regardless of any errors.

pg427 committed:
Updated compute_novelty.py

Removed local imports from files and transferred functions from the files to compute_novelty.py

pg427 committed:
Obsolete files removed

Due to code import in compute_novelty.py, removed files: known.py, molecular_similarity.py, extr_smile_molpro_by_id.py

pg427 committed:
Removed commented code

Removed commented code from compute_novelty.py

pg427 committed:
Updated compute_novelty.py

Switched error cases to include 0s instead of "NAN" ion the Dataframe.

pg427 committed:
Recency computation latency resolved

Redis is implemented in the function get_publication_info to compute recency score at a much faster rate.

pg427 committed:
Molecular similarity update

Used updated links to produce SMILEs for Molecular similarity computation in lesser time.

maximusunc commented on app/novelty/chk_gene_nmf.py:
```diff
@@ -0,0 +1,147 @@
+"""Test Clinical Evidence function."""
```
Is this file just for testing?
I really don't like the idea of having to add paths to the System PATH. Can we do this a different way?

pg427 replied:
This file was added by Marc Duby just for testing the functionality of his code. I can remove it in the next commit.
Per your comment below I'll remove system paths as well

maximusunc commented on app/novelty/chk_gene_nmf.py:
```diff
@@ -0,0 +1,147 @@
+    ]
+map_result = adapter.get_gene_nmf_novelty_for_gene_list(list_input_genes=list_gene_test, log=True)
+print(map_result)
+# def test_get_gene_nmf_novelty_for_gene_list():
```
Remove commented out code

pg427 replied:
Commented out code will be removed in the next commit

maximusunc commented on app/novelty/compute_novelty.py:
```diff
@@ -3,10 +3,22 @@
+import requests
+import redis
+
+startup_utils_path = os.path.abspath("app/novelty/gene_nmf_adapter.py")
```
Again, adding files to the system path shouldn't be needed here. These can just be imported since they're within the Appraiser "package". Please remove all system path logic.

pg427 replied:
Understood. I'll remove all system path logic.

maximusunc commented on app/novelty/compute_novelty.py:
```diff
@@ -19,79 +31,220 @@
+            nearest_neighbor_mapping.update({unknownkey: neighbors})
+        else:
+            # Calculate fingerprints for the query molecule
+            query_fp = AllChem.GetMorganFingerprint(query_mol, 2)
```
I believe these are spewing a bunch of deprecation warning log messages. See https://github.com/TranslatorSRI/answer-appraiser/issues/31

pg427 replied:
I am using an updated function (AllChem.GetMorganFingerprintAsBitVect) now to do the same. These should take care of any depreciation warnings going forward.

maximusunc commented on app/novelty/compute_novelty.py:
```diff
@@ -19,79 +31,220 @@
+    # print(smiles)
+    return smiles
+
+# async def mol_to_smile_molpro(molecules):
```
Please remove commented out code

pg427 replied:
Done. Will reflect in next commit

maximusunc commented on app/novelty/compute_novelty.py:
```diff
@@ -132,175 +286,78 @@ def recency_function_exp(number_of_publ, age_of_oldest_publ, max_number, max_age
+    for idi, i in enumerate(result['analyses']):
+        result_resource = i['resource_id']
+        edge_keys = list(i['edge_bindings'].keys())
+        if result_resource=="infores:unsecret-agent":
```
Can we find a better way to do this? What about the CQS? What about any other ARAs that get added, or ARAs that get updated and publication id locations move slightly? To me, this seems very brittle and it should be a function that looks in all places for any publication info regardless of which ARA it's from.

pg427 replied:
It was my understanding that the code to be pushed is limited to only have the 5 ARAs and not CQS. I specifically targeted the way the publication IDs are stored within each ARA to reduce computation time. Unfortunately, each ARA stores this information differently. The alternative would be to check for all locations across all ARAs which could increase computation time.

maximusunc replied:
I think the dynamic function for all ARAs far outweighs the computational overhead of checking all locations. I also think the increase in computation time will be negligible.

pg427 replied:
Got it. I can update the code to have this added in the next commit.

maximusunc commented on app/novelty/compute_novelty.py:
```diff
@@ -341,134 +397,298 @@ def result_edge_correlation(results, results_known, df):
+
+    return df_numpy
+
+# filename = 'e2ed43f8-b2c7-4f0e-9eb8-169bdca50444.json'
```
Please remove commented out code

pg427 replied:
Done. Will reflect in next commit.

maximusunc reviewed (changes requested):
- Please remove the `.idea` folder from this PR, and add it to the `.gitignore`.
- This repo also uses [black](https://pypi.org/project/black/), which is why one of the CI/CD checks is failing. This auto-formats all the code and makes sure that it is formatted to a certain standard. I would get all your other commits in order and then run this, which will make many many updates and then you can commit all these black changes together.

pg427 committed:
Novelty Score Calculation Update

Removed commented lines of code
Removed system path lines
Updated code to remove depreciation issues
Removed chk_gene_nmf.py
Updated code to look for publications in all applicable locations for all ARAs

maximusunc reviewed (changes requested):
The code no longer works "as is". All of the existing imports that have to do with gene_nmf and dcc need to be changed to relative imports.

maximusunc commented on app/novelty/compute_novelty.py:
```diff
@@ -3,95 +3,155 @@
+                "name": "id",
+                "value": i
+            })
+        response = requests.post(url, json=message)
```
We cannot use a synchronous API call here. It will block the entire worker so that it can't work on other asynchronous queries. I personally prefer `httpx`.

pg427 replied:
I've altered the code to work asynchronously using httpx instead of using requests.

pg427 committed:
Merge remote-tracking branch 'upstream/main' into novelty_score_update

pg427 committed:
Merge latest changes from upstream

pg427 committed:
Async Links and Fixed Deprecation

pg427 committed:
Added relative paths

maximusunc committed:
Update redis to use values from settings

maximusunc committed:
Turn novelty back on

maximusunc committed:
Turn off urllib3 logging

maximusunc committed:
Add some things to gitignore

maximusunc committed:
Run black

maximusunc reviewed (approved):
Looks good! Thanks for all your work @pg427 !!
